presto-main/src/main/java/com/facebook/presto/operator/PartitionedLookupSourceFactory.java

```diff
@@ -381,15 +381,18 @@ private void freePartitions()
     private class SpillAwareLookupSourceProvider
             implements LookupSourceProvider
     {
+        private LookupSource lookupSource;
+
         @Override
         public <R> R withLease(Function<LookupSourceLease, R> action)
         {
             rwLock.readLock().lock();
             try {
-                try (LookupSource lookupSource = lookupSourceSupplier.getLookupSource()) {
-                    LookupSourceLease lease = new SpillAwareLookupSourceLease(lookupSource, spillingInfo);
-                    return action.apply(lease);
+                if (lookupSource == null) {
+                    lookupSource = lookupSourceSupplier.getLookupSource();
```

Comment:
When spill epoch changes, you need to call `lookupSourceSupplier.getLS()` once again.